adoc/chapters/programming_interface.adoc

```diff
@@ -5540,358 +5423,318 @@ queue, the sub-buffer's alignment must be compatible with both the primary
+[source,role=synopsis,id=api:buffer-reinterpret]
 ----
-template <typename ReinterpretT, int ReinterpretDim>
-buffer<ReinterpretT, ReinterpretDim,
-       typename std::allocator_traits<AllocatorT>::template rebind_alloc<
-           std::remove_const_t<ReinterpretT>>>
-reinterpret(range<ReinterpretDim> reinterpretRange) const
+  template <typename ReinterpretT, int ReinterpretDim>
+  buffer<ReinterpretT, ReinterpretDim,
+         typename std::allocator_traits<AllocatorT>::template rebind_alloc<
+             ReinterpretT>>
+  reinterpret(range<ReinterpretDim> reinterpretRange) const;
 ----
-   a@ Creates and returns a reinterpreted SYCL [code]#buffer#
-      with the type specified by [code]#ReinterpretT#,
-      dimensions specified by [code]#ReinterpretDim# and range
-      specified by [code]#reinterpretRange#.  The buffer object
-      being reinterpreted can be a SYCL sub-buffer that was created
-      from a SYCL [code]#buffer# and must throw
-      [code]#exception# with the
-      [code]#errc::invalid# error code if the total
-      size in bytes represented by the type and range of the
-      reinterpreted SYCL [code]#buffer# (or sub-buffer) does not
-      equal the total size in bytes represented by the type and range
-      of this SYCL [code]#buffer# (or sub-buffer).
-      Reinterpreting a sub-buffer provides a reinterpreted view of
-      the sub-buffer only, and does not change the offset or size of
-      the sub-buffer view (in bytes) relative to the parent
-      [code]#buffer#.
+_Preconditions (2):_ Available when [code]#(ReinterpretDim == 1)# or when
+[code]#\((ReinterpretDim == Dimensions) && (sizeof(ReinterpretT) == sizeof(T)))#.
 
-a@
-[source]
-----
-template <typename ReinterpretT, int ReinterpretDim = Dimensions>
-buffer<ReinterpretT, ReinterpretDim,
-       typename std::allocator_traits<AllocatorT>::template rebind_alloc<
-           std::remove_const_t<ReinterpretT>>>
-reinterpret() const
-----
-   a@ Creates and returns a reinterpreted SYCL [code]#buffer#
-      with the type specified by [code]#ReinterpretT# and
-      dimensions specified by [code]#ReinterpretDim#.
-      Only valid when [code]#(ReinterpretDim == 1)# or when
-      [code]#\((ReinterpretDim == Dimensions) && (sizeof(ReinterpretT) == sizeof(T)))#.
-      The buffer object being reinterpreted can be a SYCL sub-buffer
-      that was created from a SYCL [code]#buffer#.  The implementation must
-      throw an [code]#exception# with the [code]#errc::invalid# error code
-      if the total size in bytes represented by this SYCL [code]#buffer# (or
-      sub-buffer) is not evenly divisible by [code]#sizeof(ReinterpretT)#.
-      Reinterpreting a sub-buffer provides a reinterpreted view
-      of the sub-buffer only,
-      and does not change the offset or size of the sub-buffer view (in bytes)
-      relative to the parent [code]#buffer#.
+_Effects:_ Creates a reinterpreted SYCL [code]#buffer#. The buffer object being
```

Comment:
I don't think we should have both an _Effects_ and a _Returns_ for this.  _Returns_ seems to be sufficient.

Reply:
I think the benefit of having Effects as well is that I could document some of the common (mostly sub-buffer) behavior.